File: phase0/ios/BoxOpPhase0/CapabilityMatrix.swift
import Foundation

/// Canonical Phase 0 capability record.
///
/// Field names and meaning are defined in
/// `phase0/shared/CAPABILITY_MATRIX.md` and
/// `phase0/shared/capability-matrix.schema.json`. Keep this struct in sync
/// with that document and with the Android equivalent
/// (`CapabilityMatrix.kt`) field-for-field.
struct CapabilityMatrix: Codable {
    let platform: String
    let osVersion: String
    let deviceModel: String
    let harnessVersion: String
    let timestamp: String

    let cameraAvailable: Bool
    let cameraAuthorizationStatus: String

    let arSupported: Bool
    let arSupportDetail: String
    let planeDetectionSupported: Bool

    let accelerometerAvailable: Bool
    let gyroscopeAvailable: Bool
    let deviceMotionAvailable: Bool

    let depthSupported: Bool
    let lidarAvailable: Bool
    let tofAvailable: Bool

    let selectedMeasurementStrategy: String
    let fallbackStrategy: String
    let knownLimitations: String
    let rawNotes: String
}

File: phase0/android/app/src/main/java/com/boxop/phase0/CapabilityMatrix.kt
package com.boxop.phase0

import org.json.JSONObject

/**
 * Canonical Phase 0 capability record.
 *
 * Field names and meaning are defined in
 * `phase0/shared/CAPABILITY_MATRIX.md` and
 * `phase0/shared/capability-matrix.schema.json`. Keep this data class in
 * sync with that document and with the iOS equivalent
 * (`CapabilityMatrix.swift`) field-for-field.
 */
data class CapabilityMatrix(
    val platform: String,
    val osVersion: String,
    val deviceModel: String,
    val harnessVersion: String,
    val timestamp: String,

    val cameraAvailable: Boolean,
    val cameraAuthorizationStatus: String,

    val arSupported: Boolean,
    val arSupportDetail: String,
    val planeDetectionSupported: Boolean,

    val accelerometerAvailable: Boolean,
    val gyroscopeAvailable: Boolean,
    val deviceMotionAvailable: Boolean,

    val depthSupported: Boolean,
    val lidarAvailable: Boolean,
    val tofAvailable: Boolean,

    val selectedMeasurementStrategy: String,
    val fallbackStrategy: String,
    val knownLimitations: String,
    val rawNotes: String
) {
    /** No JSON library dependency needed: org.json ships with Android. */
    fun toJson(): String {
        val json = JSONObject()
        json.put("platform", platform)
        json.put("osVersion", osVersion)
        json.put("deviceModel", deviceModel)
        json.put("harnessVersion", harnessVersion)
        json.put("timestamp", timestamp)
        json.put("cameraAvailable", cameraAvailable)
        json.put("cameraAuthorizationStatus", cameraAuthorizationStatus)
        json.put("arSupported", arSupported)
        json.put("arSupportDetail", arSupportDetail)
        json.put("planeDetectionSupported", planeDetectionSupported)
        json.put("accelerometerAvailable", accelerometerAvailable)
        json.put("gyroscopeAvailable", gyroscopeAvailable)
        json.put("deviceMotionAvailable", deviceMotionAvailable)
        json.put("depthSupported", depthSupported)
        json.put("lidarAvailable", lidarAvailable)
        json.put("tofAvailable", tofAvailable)
        json.put("selectedMeasurementStrategy", selectedMeasurementStrategy)
        json.put("fallbackStrategy", fallbackStrategy)
        json.put("knownLimitations", knownLimitations)
        json.put("rawNotes", rawNotes)
        return json.toString(2)
    }
}
